specs/derivation.md

```diff
@@ -473,28 +473,31 @@ updated, such that the batch-sender authentication is always accurate to the exa
 
 ### L1 Retrieval
 
-In the *L1 Retrieval* stage, we read the block we get from the outer stage (L1 traversal), and extract data from it.
-By default, the rollup operates on calldata retrieved from [batcher transactions][g-batcher-transaction] in the block,
-for each transaction:
+In the *L1 Retrieval* stage, we read the block we get from the outer stage (L1 traversal), and
+extract data from its [batcher transactions][g-batcher-transaction]. A batcher
+transaction is one with the following properties:
 
-- The receiver must be the configured batcher inbox address.
-- The sender must match the batcher address loaded from the system config matching the L1 block of the data.
+- The receiver is the configured batcher inbox address.
+- The sender is the batcher address loaded from the system config matching the L1 block of the data.
 
-Each data-transaction is versioned and contains a series of [channel frames][g-channel-frame] to be read by the
-Frame Queue, see [Batch Submission Wire Format][wire-format].
+Each batcher transaction is versioned and contains a series of [channel frames][g-channel-frame] to
+be read by the Frame Queue, see [Batch Submission Wire Format][wire-format].
+
+All non-blob (transaction-type != `0x03` (`BLOB_TX_TYPE`)) batcher transactions in the block are
+processed first; their calldata is retrieved and passed on to the next stage before proceeding.
```

Comment:
Yes understanding is correct and this is how it is currently implemented.     I didn't think we could guarantee a total order of batches in the L1 since that's up to the block builders?

Reply:
I don't really understand what you mean. The total ordering of L2 blocks should be determined by the order in which batches are included in L1. Why is it better to have this out of order rule in consensus? Seems like unnecessary complexity, ie yet another unit test that needs to be written. I think that it should be as simple as the order of batches included in L1 across calldata and blobs should match the total ordering of batches perceived by consensus

Reply:
Possibly I was talking about something else - what I meant is that the order in which our batcher transactions appear in the block is not something we control, it's up to the block builders.  But once built of course the ordering is then forever fixed. For derivation purposes we only need a canonical ordering in which to process the batches as they appear in the block.  We went with calldata first then blobdata second (by order of appearance in the block) since this is easier to implement and (to me) easier to explain in the spec: you have a separate for each loop for each type,  and don't otherwise need to condition on tx-type. (there are different rules for handling each since blobs involve downloading & further validation).

We can certainly change it to instead have calldata & blobdata interleaved according to the order in which their respective transactions appear. I don't personally find that any simpler, but I think this is somewhat subjective.

Reply:
>  the order in which our batcher transactions appear in the block is not something we control

Nonces must increment so we do control this. We don't control the exact block that the tx is included in but we do control the ordering in which our transactions are included because we set the nonce field.

From the pov of the batcher, not interleaving them would make dynamic strategies very difficult to implement. People may think that the dynamic strategies will be rare but I would bet that things like ordinals will push blob fees higher than calldata at times of high demand.

Definitely open to hearing the opinions of others but breaking the invariant that the total ordering of batches in L1 match to the total ordering of deserialized batches adds complexity to the protocol

Reply:
Ack. proto chimed in below and is aligned with preserving tx order so will make the change and ping when done.

Reply:
OK change is now in the spec here, forgot the implementation is another PR, will make that change next.